SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-3x/tsWxGTUI_Py3x/ManPages/tsManPagesLibGUI/tsWxApp.man

Help on module tsWxApp:

NAME
    tsWxApp

FILE
    ./tsLibGUI/tsWxPkg/src/tsWxApp.py

DESCRIPTION
    tsWxApp.py - Class to represent the application. It is used to
    bootstrap the wxPython system and initialize the underlying gui
    toolkit. It sets and gets application-wide properties. It
    implements the windowing system main message or event loop. It
    dispatches events to window instances.

CLASSES
    __builtin__.object
        tsLibCLI.tsReportUtilityPkg.src.tsReportUtilities.TsReportUtilities
        tsCustomStdioFile
    tsWxPyApp.PyApp(tsWxEvtHandler.EvtHandler)
        App
    
    class App(tsWxPyApp.PyApp)
     |  The wx.App class represents the application and is used to:
     |  
     |      * bootstrap the wxPython system and initialize the underlying gui
     |        toolkit
     |      * set and get application-wide properties
     |      * implement the windowing system main message or event loop, and to
     |        dispatch events to window instances
     |      * etc.
     |  
     |  Every application must have a wx.App instance, and all creation of UI
     |  objects should be delayed until after the wx.App object has been created
     |  in order to ensure that the gui platform and wxWidgets have been fully
     |  initialized.
     |  
     |  Normally you would derive from this class and implement an OnInit method
     |  that creates a frame and then calls self.SetTopWindow(frame).
     |  
     |  Method resolution order:
     |      App
     |      tsWxPyApp.PyApp
     |      tsWxEvtHandler.EvtHandler
     |      tsWxObject.Object
     |      __builtin__.object
     |  
     |  Methods defined here:
     |  
     |  Destroy(self)
     |  
     |  GetMainLoop(self)
     |      Return the main GUI event loop instance.
     |      
     |      Note: Overrides tsWxPyApp.MainLoop.
     |  
     |  MainLoop(self)
     |      Execute the main GUI event loop.
     |      
     |      Note: Overrides tsWxPyApp.MainLoop.
     |  
     |  OnInit(self)
     |      This must be provided by the application, and will usually create the
     |      application main window, optionally calling {wxApp::SetTopWindow}.
     |      You may use {OnExit} to clean up anything initialized here, provided
     |      that the function returns true.
     |      
     |      Notice that if you want to to use the command line processing provided
     |      by wxWidgets you have to call the base class version in the derived
     |      class OnInit().
     |      
     |      Return true to continue processing, false to exit the application
     |      immediately.
     |      
     |      Note: This module does not appear in wPython API for wxApp or wxPyApp.
     |  
     |  OnPreInit(self)
     |      Things that must be done after _BootstrapApp has done its thing, but
     |      would be nice if they were already done by the time that OnInit is
     |      called.
     |  
     |  RedirectStdio(self, filename=None)
     |      Redirect sys.stdout and sys.stderr to a file or a popup window.
     |  
     |  RestoreStdio(self)
     |      Restore sys.stdout and sys.stderr.
     |  
     |  SetOutputWindowAttributes(self, title=None, pos=None, size=None)
     |      Set the title, position and/or size of the output window if the
     |      stdio has been redirected.
     |  
     |  SetTopWindow(self, frame)
     |      Set the main top level window.
     |  
     |  __init__(self, redirect=False, filename=None, useBestVisual=False, clearSigInt=True)
     |      Construct a wx.App object.
     |      
     |      Parameters:
     |      
     |          redirect - Should sys.stdout and sys.stderr be redirected?
     |          Defaults to True on Windows and Mac, False otherwise. If filename
     |          is None then output will be redirected to a window that pops up
     |          as needed. (You can control what kind of window is created for
     |          the output by resetting the class variable outputWindowClass to
     |          a class of your choosing.)
     |      
     |          filename - The name of a file to redirect output to, if redirect
     |          is True.
     |      
     |          useBestVisual - Should the app try to use the best available
     |          visual provided by the system (only relevant on systems that have
     |          more than one visual.) This parameter must be used instead of
     |          calling SetUseBestVisual later on because it must be set before
     |          the underlying GUI toolkit is initialized.
     |      
     |          clearSigInt - Should SIGINT be cleared? This allows the app to
     |          terminate upon a Ctrl-C in the console like other GUI apps will.
     |      
     |      Note: You should override OnInit to do applicaition initialization
     |      to ensure that the system, toolkit and wxWidgets are fully initialized.
     |  
     |  tsBootstrapApp(self)
     |      Initialize the native GUI (curses) screen. Perform any application
     |      specified pre-initialization and initialization activities.
     |  
     |  tsSetStdioWindowAttributes(self, title=None, pos=None, size=None, style=None, name=None)
     |      Set the title, position, size, style and name of the output window
     |      if the stdio has been redirected.
     |  
     |  ----------------------------------------------------------------------
     |  Data and other attributes defined here:
     |  
     |  outputWindowClass = <class 'tsWxPyOnDemandOutputWindow.PyOnDemandOutpu...
     |      A class that can be used for redirecting Python stdout and stderr streams.
     |      It will do nothing until something is written to the stream at which point
     |      it will create a Frame with a text area and write the text there.
     |  
     |  wxTheApp = None
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from tsWxPyApp.PyApp:
     |  
     |  Dispatch(self)
     |      Process the first event in the event queue (blocks until an event
     |      appears if there are none currently).
     |  
     |  Exit(self)
     |      Exit the main loop thus terminating the application. :see: wx.Exit
     |  
     |  ExitMainLoop(self)
     |      Exit the main GUI loop during the next iteration of the main loop,
     |      (i.e. it does not stop the program immediately!)
     |      (i.e....).
     |  
     |  FilterEvent(self, event)
     |  
     |  GetAppName(self)
     |      Get the application name.
     |  
     |  GetAssertMode(self)
     |      Get the current OnAssert behaviour setting.
     |  
     |  GetClassName(self)
     |      Get the application class name.
     |  
     |  GetExitOnFrameDelete(self)
     |      Get the current exit behaviour setting.
     |  
     |  GetLayoutDirection(self)
     |      Return the layout direction for the current locale.
     |  
     |  GetPrintMode(self)
     |      Get the current Print Mode behaviour setting.
     |  
     |  GetTopWindow(self)
     |      Return the main top level window (if it had not been set previously
     |      with SetTopWindow(), will return just some top level window and, if
     |      there not any, will return None)
     |  
     |  GetTraits(self)
     |      Return (and create if necessary) the app traits object to which we
     |      delegate for everything which either should be configurable by the
     |      user (then he can change the default behaviour simply by overriding
     |      CreateTraits() and returning his own traits object) or which is
     |      GUI/console dependent as then wx.AppTraits allows us to abstract
     |      the differences behind the common facade.
     |  
     |  GetUseBestVisual(self)
     |      Get current UseBestVisual setting.
     |  
     |  GetVendorName(self)
     |      Get the application vendor name.
     |  
     |  HandleEvent(self, handler, funct, event)
     |      This function simply invokes the given method funct of the
     |      specified event handler, handler, with the event as parameter.
     |  
     |  IsActive(self)
     |      Return True if our app has focus.
     |  
     |  OnExit(self)
     |      Override this member function for any processing which needs to be
     |      done as the application is about to exit. OnExit is called after
     |      destroying all application windows and controls, but before wxWidgets
     |      cleanup. Note that it is not called at all if {OnInit} failed.
     |      
     |      The return value of this function is currently ignored, return the
     |      same value as returned by the base class method if you override it.
     |  
     |  Pending(self)
     |      Returns True if there are unprocessed events in the event queue.
     |  
     |  ProcessIdle(self)
     |      Called from the MainLoop when the application becomes idle (there
     |      are no pending events) and sends a wx.IdleEvent to all interested
     |      parties. Returns True if more idle events are needed, False if not.
     |  
     |  ProcessIdleEvent(self)
     |      Called from the MainLoop when the application becomes idle (there
     |      are no pending events) and sends a wx.IdleEvent to all interested
     |      parties. Returns True if more idle events are needed, False if not.
     |  
     |  ProcessIdleNoEvent(self)
     |      Called from the MainLoop when the application becomes idle (there
     |      are no pending events) and sends a wx.IdleEvent to all interested
     |      parties. Returns True if more idle events are needed, False if not.
     |  
     |  ProcessPendingEvents(self)
     |      Process all events in the Pending Events list -- it is necessary
     |      to call this function to process posted events. This normally happens
     |      during each event loop iteration.
     |  
     |  SendIdleEvents(self, win, event)
     |      Send idle event to window and all subwindows.Returns True if more
     |      idle time is requested.
     |  
     |  SetAppName(self, name)
     |      Set the application name. This value may be used automatically by
     |      wx.Config and such.
     |  
     |  SetAssertMode(self, mode)
     |      Set the OnAssert behaviour for debug and hybrid builds. The
     |      following flags may be or-ed together:

Documents/NEWS.txt
#-----------------------------------------------------------
#"Time-stamp: <07/08/2015  4:48:41 AM rsg>"
#-----------------------------------------------------------

========================= NEWS.txt =========================

   +----+----+  TeamSTARS "tsWxGTUI_PyVx" Toolkit
   | ts | Wx |      with Python 2x & Python 3x based
   +----+----+         Command Line Interface (CLI)
   | G T U I |      and "Curses"-based "wxPython"-style, 
   +---------+         Graphical-Text User Interface (GUI)
 
   Get that cross-platform, pixel-mode "wxPython" feeling
   on character-mode 8-/16-color (xterm-family) & non-color
   (vt100-family) terminals and terminal emulators.

   You can find this and other plain-text files in the
   Toolkit subdirectory named:

       "./<Toolkit Recipient's Repository>/Documents".

       <Your Working Repository>
       (e.g. "tsWxGTUI_PyVx_Repository") 
         |
         |
         +-- ["Documents"]

=========================== News ===========================

The pre-alpha release of the TeamSTARS "tsWxGTUI_PyVx-0.0.0"
Toolkit is now available. It includes the following two
versions abd their associated variants (Python Distribution
Utilities ("Distutils") installable "Site-Packages" and non-
installable "Developer-Sandboxes" for enhancement and bug
fix development and experimentation without interference
with the already installed Site-Packages):

1. The TeamSTARS "tsWxGTUI_Py2x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   second generation Python 2.x programming language which
   is widely available.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. The final 2.x
   version 2.7 release came out in mid-2010, with a state-
   ment of extended support for this end-of-life release.
   The 2.x branch will see no new major releases after that.

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 2.0.0-2.7.9.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 2.6.4-2.7.9.

2. The TeamSTARS "tsWxGTUI_Py3x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   third generation Python 3.x programming language which
   is growing in availability.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. 3.x is under
   active development and has already seen over five years
   of stable releases, including version 3.3 in 2012 and
   3.4 in 2014. This means that all recent standard library
   improvements, for example, are only available by default
   in Python 3.x.

   Guido van Rossum (the original creator of the Python
   language) decided to clean up Python 2.x properly,
   with less regard for backwards compatibility than is
   the case for new releases in the 2.x range. The most
   drastic improvement is the better Unicode support
   (with all text strings being Unicode by default) as
   well as saner bytes/Unicode separation.

   Besides, several aspects of the core language (such as
   print and exec being statements, integers using floor
   division) have been adjusted to be easier for newcomers
   to learn and to be more consistent with the rest of the
   language, and old cruft has been removed (for example,
   all classes are now new-style, "range()" returns a memory
   efficient iterable, not a list as in 2.x).

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 3.0.0-3.4.2.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 3.1.5-3.4.2.

======================= End-Of-File ========================


SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-3x/tsWxGTUI_Py3x/tsLibGUI/tsWxPkg/src/tsWxApp.py
#! /usr/bin/env python
# "Time-stamp: <10/25/2013 11:04:28 AM rsg>"
'''
tsWxApp.py - Class to represent the application. It is used to
bootstrap the wxPython system and initialize the underlying gui
toolkit. It sets and gets application-wide properties. It
implements the windowing system main message or event loop. It
dispatches events to window instances.
'''
#################################################################
#
# File: tsWxApp.py
#
# Purpose:
#
#    Class to represent the application. It is used to bootstrap
#    the wxPython system and initialize the underlying gui
#    toolkit. It sets and gets application-wide properties. It
#    implements the windowing system main message or event loop.
#    It dispatches events to window instances.
#
# Usage (example):
#
#    # Import
#
#    from tsWxApp import App
#
# Requirements:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Capabilities:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Limitations:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Notes:
#
#    None
#
# Methods:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Classes:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Modifications:
#
#    2010/11/01 rsg Added logic to initialize App.wxTheApp.
#
# ToDo:
#
#    None.
#
#################################################################

__title__     = 'tsWxApp'
__version__   = '1.0.6'
__date__      = '04/01/2013'
__authors__   = 'Richard S. Gordon'
__copyright__ = 'Copyright (c) 2007-2013 ' + \
                '%s.\n\t\tAll rights reserved.' % __authors__
__license__   = 'GNU General Public License, ' + \
                'Version 3, 29 June 2007'
__credits__   = '\n\n  Credits: ' + \
                '\n\n\t  tsLibGUI Import & Application Launch Features: ' + \
                '\n\t  Copyright (c) 2007-2009 Frederick A. Kier.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\n\t  Python Curses Module API & ' + \
                'Run Time Library Features:' + \
                '\n\t  Copyright (c) 2001-2013 ' +\
                'Python Software Foundation.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  PSF License Agreement for Python 2.7.3 & 3.3.0' + \
                '\n\n\t  wxWidgets (formerly wxWindows) & ' + \
                'wxPython API Features:' + \
                '\n\t  Copyright (c) 1992-2008 Julian Smart, ' + \
                'Robert Roebling,' + \
                '\n\t\t\tVadim Zeitlin and other members of the ' + \
                '\n\t\t\twxWidgets team.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  wxWindows Library License' + \
                '\n\n\t  nCurses API & Run Time Library Features:' + \
                '\n\t  Copyright (c) 1998-2011 ' + \
                'Free Software Foundation, Inc.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  GNU General Public License, ' + \
                'Version 3, 29 June 2007'

__line1__ = '%s, v%s (build %s)' % (__title__, __version__, __date__)
__line2__ = 'Author(s): %s' % __authors__
__line3__ = '%s' % __copyright__

if len( __credits__) == 0:
    __line4__ = '%s' % __license__
else:
    __line4__ = '%s%s' % (__license__, __credits__)

__header__ = '\n\n%s\n\n  %s\n  %s\n  %s\n' % (__line1__,
                                               __line2__,
                                               __line3__,
                                               __line4__)

mainTitleVersionDate = __line1__

import sys
import signal
import time

import tsExceptions as tse
from tsReportUtilities import TsReportUtilities as tsru

import tsWxGlobals as wx
from tsWxPyOnDemandOutputWindow import PyOnDemandOutputWindow as wxPyOnDemandOutputWindow
from tsWxPyApp import PyApp
from tsWxPoint import Point as wxPoint
from tsWxSize import Size as wxSize

#---------------------------------------------------------------------------

##DEBUG = True
DEBUG = wx.Debug_GUI_Launch | \
        wx.Debug_GUI_Progress | \
        wx.Debug_GUI_Termination | \
        wx.Debug_GUI_Exceptions

##VERBOSE = True
VERBOSE = wx.Debug_GUI_Configuration

#---------------------------------------------------------------------------

class App(PyApp):
    '''
    The wx.App class represents the application and is used to:

        * bootstrap the wxPython system and initialize the underlying gui
          toolkit
        * set and get application-wide properties
        * implement the windowing system main message or event loop, and to
          dispatch events to window instances
        * etc.

    Every application must have a wx.App instance, and all creation of UI
    objects should be delayed until after the wx.App object has been created
    in order to ensure that the gui platform and wxWidgets have been fully
    initialized.

    Normally you would derive from this class and implement an OnInit method
    that creates a frame and then calls self.SetTopWindow(frame).
    '''
    # Class Variables:
    outputWindowClass = wxPyOnDemandOutputWindow
    wxTheApp = None

    #-----------------------------------------------------------------------

    def __init__(self,
                 redirect=False,
                 filename=None,
                 useBestVisual=False,
                 clearSigInt=True):
        '''
        Construct a wx.App object.

        Parameters:

            redirect - Should sys.stdout and sys.stderr be redirected?
            Defaults to True on Windows and Mac, False otherwise. If filename
            is None then output will be redirected to a window that pops up
            as needed. (You can control what kind of window is created for
            the output by resetting the class variable outputWindowClass to
            a class of your choosing.)

            filename - The name of a file to redirect output to, if redirect
            is True.

            useBestVisual - Should the app try to use the best available
            visual provided by the system (only relevant on systems that have
            more than one visual.) This parameter must be used instead of
            calling SetUseBestVisual later on because it must be set before
            the underlying GUI toolkit is initialized.

            clearSigInt - Should SIGINT be cleared? This allows the app to
            terminate upon a Ctrl-C in the console like other GUI apps will.

        Note: You should override OnInit to do applicaition initialization
        to ensure that the system, toolkit and wxWidgets are fully initialized.
        '''
        theClass = 'App'

        wx.RegisterFirstCallerClassName(self, theClass)

        # Capture initial caller parametsrs before they are changed
        self.ts_redirect = redirect
        self.ts_filename = filename
        self.ts_useBestVisual = useBestVisual
        self.ts_clearSigInt = clearSigInt
        self.outputWindowAttributes = None
        self.stdioWin = None
        self.stdioLog = None

        self.ts_ReserveRedirectArea = redirect and \
                                      filename is None

        self.ts_ReserveTaskBarArea = wx.ThemeToUse['TaskBar']['Enable']

        PyApp.__init__(self)

        self.tsBeginClassRegistration(theClass, id)

        if App.wxTheApp is None:

            App.wxTheApp = self
 
        # Make sure we can create a GUI
        if not self.IsDisplayAvailable():

            msg = "Unable to create GUI"
            raise tse.UserInterfaceException(
                tse.CHARACTER_GRAPHICS_NOT_AVAILABLE, msg)
 
        # This has to be done before OnInit
        self.SetUseBestVisual(useBestVisual)

        # Set the default handler for SIGINT.  This fixes a problem
        # where if Ctrl-C is pressed in the console that started this
        # app then it will not appear to do anything, (not even send
        # KeyboardInterrupt???)  but will later segfault on exit.  By
        # setting the default handler then the app will exit, as
        # expected (depending on platform.)
        if clearSigInt:
            try:
                signal.signal(signal.SIGINT, signal.SIG_DFL)
            except ImportError:
                pass

        # Save and redirect the stdio to a window?
        self.stdioWin = None
        self.saveStdio = (sys.stdout, sys.stderr)

##        if redirect:
##            self.RedirectStdio(self.ts_filename)

        try:
            # Initialize the specified stdout/stderr redirection.
            if self.ts_redirect:
                self.logger.debug(
                    '    Redirection: (%s); Filename: %s.' % (
                        self.ts_redirect,
                        self.ts_filename))
                self.RedirectStdio(filename=self.ts_filename)
                self.logger.debug('  Performed Redirection.')

        except AttributeError as RedirectError:
            self.logger.error('  Skipped Redirection. %s' % RedirectError)
            raise

        # Use Python's install prefix as the default
##        wx.StandardPaths.Get().SetInstallPrefix(sys.prefix)

        # Until the new native control for wxMac is up to par, still use
        # the generic one.
##        wx.SystemOptions.SetOptionInt("mac.listctrl.always_use_generic", 1)

        # This finishes the initialization of wxWindows and then calls
        # the OnInit that should be present in the derived class
        self.tsBootstrapApp()

        self.tsEndClassRegistration(theClass)

    #-----------------------------------------------------------------------

##    def __del__(self, destroy):
##        '''
##        '''
##        if destroy:
##            self.Destroy()

    #-----------------------------------------------------------------------

    def Destroy(self):
        '''
        '''
        if self.ts_redirect:
            self.RestoreStdio()

    #-----------------------------------------------------------------------

    def GetMainLoop(self):
        '''
        Return the main GUI event loop instance.

        Note: Overrides tsWxPyApp.MainLoop.
        '''
        return (PyApp.MainLoop(self))

    #-----------------------------------------------------------------------

    def MainLoop(self):
        '''
        Execute the main GUI event loop.

        Note: Overrides tsWxPyApp.MainLoop.
        '''
        PyApp.MainLoop(self)
        self.RestoreStdio()

    #-----------------------------------------------------------------------

    def OnPreInit(self):
        '''
        Things that must be done after _BootstrapApp has done its thing, but
        would be nice if they were already done by the time that OnInit is
        called.
        '''
        status = PyApp.OnPreInit(self)
        # Force redirection to start before application Frame(s).
        if status:
            print('')
        return (status)

    #-----------------------------------------------------------------------

    def RedirectStdio(self, filename=None):
        '''
        Redirect sys.stdout and sys.stderr to a file or a popup window.
        '''
        # Verify user accessibility of one library known to be in hierarchy.
        try:
            preRedirectStdioDevices = {'stdout': sys.stdout,
                                       'stderr': sys.stderr,
                                       ' stdin': sys.stdin}

            for theKey in list(preRedirectStdioDevices.keys()):
                self.logger.debug(
                    '    Saved %s %s' % (
                        theKey, preRedirectStdioDevices[theKey]))
        except Exception as preRedirectStdioDevicesError:
            msg = "tsWxApp: %s" % preRedirectStdioDevicesError
            raise tse.ProgramException(
                tse.APPLICATION_TRAP, msg)

        # Save configuration for future restoration by RestoreStdio.
        self.saveStdio = (sys.stdout, sys.stderr)

        if filename is None:

            # Redirect output to a window on the screen.
            # Capture redirected output to a default file for scrolling.
            # TBD - How can this support the ThemeToUse Timestamp feature?
            self.stdioWin = self.outputWindowClass(wx.ThemeToUse['Stdio']['Title'])
            sys.stdout = self.stdioWin
            sys.stderr = self.stdioWin

            try:
                windowStdioDevices = {'stdout': sys.stdout,
                                      'stderr': sys.stderr,
                                      ' stdin': sys.stdin}

                for theKey in list(windowStdioDevices.keys()):
                    self.logger.debug(
                        '    Saved %s %s' % (
                            theKey, windowStdioDevices[theKey]))
            except Exception as windowStdioDevicesError:
                msg = "tsWxApp: %s" % windowStdioDevicesError
                raise tse.ProgramException(
                    tse.APPLICATION_TRAP, msg)

        else:
 
            # Redirect output to the specified file.
            # Setting buffer size of 0 eliminates need for flushing.
            if wx.ThemeToUse['Stdio']['Timestamp']:
                self.stdioLog = tsCustomStdioFile(filename, 'w', 1)
            else:
                self.stdioLog = open(filename, 'w', 1)

            sys.stdout = self.stdioLog
            sys.stderr = self.stdioLog

            try:
                fileStdioDevices = {'stdout': sys.stdout,
                                      'stderr': sys.stderr,
                                      ' stdin': sys.stdin}

                for theKey in list(fileStdioDevices.keys()):
                    self.logger.debug(
                        '    Saved %s %s' % (
                            theKey, fileStdioDevices[theKey]))
            except Exception as fileStdioDevicesError:
                msg = "tsWxApp: %s" % fileStdioDevicesError
                raise tse.ProgramException(
                    tse.APPLICATION_TRAP, msg)

            theLogFileHeader = tsru.getSeparatorString(
                title='Begin %s on %s' % (
                    'PRINT/STDOUT/STDERR log',
                    tsru.getDateAndTimeString(time.time())),
                indent=0,
                position=tsru.layout['TitleIndent'],
                separatorCharacter='$',
                tab=4)

            self.stdioLog.write('%s\n\n' % theLogFileHeader)
            self.stdioLog.write('%s - Started logging to file "%s".\n\n' % (
                tsru.getDateTimeString(time.time()),
                filename))

            msg1 = 'Print statements and other standard output '
            msg2 = 'will now be directed to this file.'
            self.stdioLog.write('%s\n' % (msg1 + msg2))

            self.stdioLog.flush()

    #-----------------------------------------------------------------------

    def RestoreStdio(self):
        '''
        Restore sys.stdout and sys.stderr.
        '''
        try:
            (sys.stdout, sys.stderr) = self.saveStdio
        except AttributeError:
            pass

        self.stdioLog = None
        self.stdioWin = None

    #-----------------------------------------------------------------------

    def SetOutputWindowAttributes(self,
                                  title=None,
                                  pos=None,
                                  size=None):
        '''
        Set the title, position and/or size of the output window if the
        stdio has been redirected.
        '''
        if self.stdioWin:
            # TBD - Cannot set attributes when window does not exist.
            if title is not None:
                self.stdioWin.title = title

            if pos is not None:
                self.stdioWin.pos = pos

            if size is not None:
                self.stdioWin.size = size

    #-----------------------------------------------------------------------

    def SetTopWindow(self, frame):
        '''
        Set the main top level window.
        '''
        self.ts_TopWindow = frame

        if self.stdioWin:
            self.stdioWin.SetParent(frame)
        PyApp.SetTopWindow(self, frame)

    #-----------------------------------------------------------------------
    # Begin tsWx API Extensions

    def OnInit(self):
        '''
        This must be provided by the application, and will usually create the
        application main window, optionally calling {wxApp::SetTopWindow}.
        You may use {OnExit} to clean up anything initialized here, provided
        that the function returns true.

        Notice that if you want to to use the command line processing provided
        by wxWidgets you have to call the base class version in the derived
        class OnInit().

        Return true to continue processing, false to exit the application
        immediately.

        Note: This module does not appear in wPython API for wxApp or wxPyApp.
        '''
        # TBD - Should keepgoing if method is not provided by Application.
        keepgoing = True
        if keepgoing:
            msg = "tsWxApp: Skipped by OnInit stub"
            self.logger.debug(msg)
        else:
            msg = "tsWxApp: Aborted by OnInit stub"
            raise tse.ProgramException(
                tse.APPLICATION_TRAP, msg)

        return (keepgoing)

    #-----------------------------------------------------------------------

    def tsBootstrapApp(self):
        '''
        Initialize the native GUI (curses) screen. Perform any application
        specified pre-initialization and initialization activities.
        '''
        self.logger.debug('  Begin BootstrapApp.')

        try:
            # Perform any application specified pre-initialization activities
            # that must be done after _BootstrapApp has done its thing, but
            # would be nice if they were already done by the time that OnInit
            # is called.
            self.OnPreInit()
            self.logger.debug('  Performed OnPreInit.')

        except AttributeError as OnPreInitError:
            self.logger.error('  Skipped OnPreInit. %s' % OnPreInitError)
            msg = 'CheckOnPreInitError: %s' % OnPreInitError
            raise tse.ProgramException(tse.APPLICATION_TRAP, msg)

        try:
            # Perform any application specified outer-loop initialization
            # activities such as the creation of frames..
            self.OnInit()
            self.logger.debug('  Performed OnInit.')

        except AttributeError as OnInitError:
            self.logger.error('  Skipped OnInit. %s' % OnInitError)
            msg = 'CheckOnInitError: %s' % OnInitError
            raise tse.ProgramException(tse.APPLICATION_TRAP, msg)

        self.logger.debug('  End BootstrapApp.')

    #-----------------------------------------------------------------------

    def tsSetStdioWindowAttributes(self,
                                   title=None,
                                   pos=None,
                                   size=None,
                                   style=None,
                                   name=None):
        '''
        Set the title, position, size, style and name of the output window
        if the stdio has been redirected.
        '''
        if self.stdioWin:
            # TBD - Cannot set attributes when window does not exist.
            myRect = self.Display.theRedirectedStdioArea
            if title is not None:
                self.stdioWin.title = title
            else:
                self.stdioWin.title = wx.ThemeToUse['Stdio']['Title']

            if pos is not None:
                self.stdioWin.pos = pos
            else:
                self.stdioWin.pos = wxPoint(myRect.x, myRect.y)

            if size is not None:
                self.stdioWin.size = size
            else:
                self.stdioWin.size = wxSize(myRect.width, myRect.height)

            if style is not None:
                self.stdioWin.style = style
            elif wx.ThemeToUse['Stdio']['Title'] == wx.StdioRedirectedTitleStr:
                self.stdioWin.style = wx.DEFAULT_STDIO_STYLE
            else:
                self.stdioWin.style = wx.DEFAULT_FRAME_STYLE

            if name is not None:
                self.stdioWin.name = name
            elif wx.ThemeToUse['Stdio']['Title'] == wx.StdioRedirectedTitleStr:
                self.stdioWin.name = wx.StdioNameStr
            else:
                self.stdioWin.name = wx.FrameNameStr

    # End tsWx API Extensions
    #-----------------------------------------------------------------------

class tsCustomStdioFile(object):
    '''
    Class to open the specified file and then prefix output with a
    system timestamp.
    '''
    def __init__(self, name, mode, buffer):
        '''
        Construct class.
        '''
        self.ts_LineCount = 0
        self.ts_LastHeaderLine = 7
        self.ts_File = open(name, mode, buffer)

    #-----------------------------------------------------------------------

    def write(self, text):
        '''
        Output specified text with the appropriate timestamp
        prefix.
        '''
        self.ts_LineCount += 1
        if self.ts_LineCount <= self.ts_LastHeaderLine or \
           text == '\n':

            self.ts_File.write(text)

        else:

            self.ts_File.write('%s - %s' % (
                tsru.getDateTimeString(time.time()), text))

#---------------------------------------------------------------------------

if __name__ == '__main__':

    print(__header__)
